### DISCUSSION TITLE
How to handle IP change of a peer

### DISCUSSION THREAD

**Discussion by StefanoSetti** [libp2p/rust-libp2p#5463] [Q&A] [answered]
In my project, written in Rust, I have a network composed by multiple peers that shares the same code.

Each peer has this behavior:

```rust
#[derive(NetworkBehaviour)]
pub struct Behaviour {
    pub autonat: autonat::Behaviour,
    pub gossipsub: gossipsub::Behaviour,
    pub identify: identify::Behaviour,
    pub kademlia: kad::Behaviour<MemoryStore>,
    pub reqres: request_response::Behaviour<ExtComm>,
}
```

One of the group's peer is in a private networks, it correctly connects to the bootstrap peer and it behaves correctly. Once in a while it logs multiple `SwarmEvent::ConnectionClosed` and it stop behaving unexpectedly.

 It seems like its router changes the public IP and the other peers can't connect to it anymore.
 
 Do you have any suggestions to help me on this?

**Comment by dariusc93** [comment]
Are you using an unspecified address (eg `0.0.0.0`) or are you listening to a specific ip for the peers? While I never personally dealt with address changes in libp2p, my assumption is if its the former and there is a report of a new ip and it should begin listening on it once it is reported from the interface, but it does make sense that the connection would close because the previous address used is no longer route-able. In that case the peer should begin connecting to the bootstrap peer(s), announcing its presence in the network and attempting to reestablish connections to other peers.

**Reply by StefanoSetti** [reply]
I'm using an unspecified address...so for what I can understand, this task is done by the libp2p itself automatically, right?

**Reply by dariusc93** [reply, marked as answer]
From previous testing, yes, however as I reread your OP, I am not sure if libp2p deals with dynamic IP. Assuming it is reported to the interface it should, but this *might* require some configuring of the dhcp client on your end. Regardless, if it can be picked up by [if-watch](https://github.com/mxinden/if-watch) (which is what `libp2p-{tcp,mdns,quic}` use) then i can assume it will be handled internally, although for the connections, i would have to check to see how `AddressChange` is emitted under the existing connection.
